src/os/src/os_log.c

```diff
@@ -0,0 +1,305 @@
+    _In_ log_sink_t *sink,
+    _In_opt_ dds_log_write_fn_t func,
+    _In_opt_ void *ptr)
+{
+    assert(sink != NULL);
+
+    /* No life cycle management is done for log sinks, the caller is
+       responsible for that. Ensure this operation is deterministic and that on
+       return, no thread in the DDS stack still uses the deprecated sink. */
+    lock_sink(WRLOCK);
+
+    if (func == NULL) {
+        sink->func = NULL;
+        sink->ptr = NULL;
+    } else {
+        sink->func = func;
+        sink->ptr = ptr;
+    }
+
+    unlock_sink();
+}
+
+/* dds_set_log_file must be considered private and must not appear in os_log.h.
+   The only other location the function is declared is in q_config.c, users
+   must not call this function. The same is true for dds_set_trace_file. */
+void dds_set_log_file(_In_ FILE *file)
+{
+    os_once(&lock_inited, init_lock);
+    os_rwlockWrite(&lock);
+    log_file = file;
+    os_rwlockUnlock(&lock);
+}
+
+void dds_set_trace_file(_In_ FILE *file)
+{
+    os_once(&lock_inited, init_lock);
+    os_rwlockWrite(&lock);
+    trace_file = file;
+    os_rwlockUnlock(&lock);
+}
+
+void dds_set_log_sink(
+    _In_opt_ dds_log_write_fn_t callback,
+    _In_opt_ void *userdata)
+{
+    set_log_sink(&log_sink, callback, userdata);
+}
+
+void dds_set_trace_sink(
+    _In_opt_ dds_log_write_fn_t callback,
+    _In_opt_ void *userdata)
+{
+    set_log_sink(&trace_sink, callback, userdata);
+}
+
+extern inline uint32_t
+dds_get_log_mask(void);
+
+void
+dds_set_log_mask(_In_ uint32_t cats)
+{
+    *dds_log_mask = (cats & (DDS_LOG_MASK | DDS_TRACE_MASK));
+}
+
+#define TRACE (1<<0)
+#define LOG (1<<1)
+#define CREATE_STRUCT (1<<2)
+#define PRINT_HEADER (1<<3)
+
+static void vlog(
+    _In_ uint32_t cat,
+    _In_z_ const char *file,
+    _In_ uint32_t line,
+    _In_z_ const char *func,
+    _In_z_ const char *fmt,
+    va_list ap)
+{
+    int n, trunc = 0;
+    size_t nrem, off = BUF_OFFSET;
+    log_buffer_t *lb;
+    dds_log_data_t data;
+
+    if (*fmt == 0) {
+        return;
+    }
+
+    lock_sink(RDLOCK);
+    lb = &log_buffer;
+
+    /* Thread-local buffer is always initialized with all zeroes. The pos
+       member must always be greater or equal to BUF_OFFSET. */
+    if (lb->pos < BUF_OFFSET) {
+        lb->pos = BUF_OFFSET;
+        lb->buf[lb->pos] = 0;
+    }
+    nrem = lb->bufsz - lb->pos;
+    if (nrem > 0) {
+        n = os_vsnprintf(lb->buf + lb->pos, nrem, fmt, ap);
+        if (n >= 0 && (size_t) n < nrem) {
+            lb->pos += (size_t) n;
+        } else {
+            lb->pos += nrem;
+            trunc = 1;
+        }
+        if (trunc) {
+            static const char msg[] = "(trunc)\n";
+            const size_t msglen = sizeof (msg) - 1;
+            assert(lb->pos <= lb->bufsz);
+            assert(lb->pos >= msglen);
+            memcpy(lb->buf + lb->pos - msglen, msg, msglen);
+        }
+    }
+
+    /* Trace messages are delivered to the trace sink, non-trace messages are
```

Comment:
(2) Can't do that because whether or not a message should end up in the trace, log or both files/sinks, is determined by the enabled categories, which are set independently from the sink.